python/ray/tune/examples/mnist_pytorch_lightning.py

```diff
@@ -0,0 +1,252 @@
+        avg_acc = torch.stack([x["val_accuracy"] for x in outputs]).mean()
+        tensorboard_logs = {"val_loss": avg_loss, "val_accuracy": avg_acc}
+
+        return {
+            "avg_val_loss": avg_loss,
+            "avg_val_accuracy": avg_acc,
+            "log": tensorboard_logs
+        }
+
+    @staticmethod
+    def download_data(data_dir):
+        transform = transforms.Compose([
+            transforms.ToTensor(),
+            transforms.Normalize((0.1307, ), (0.3081, ))
+        ])
+        return MNIST(data_dir, train=True, download=True, transform=transform)
+
+    def prepare_data(self):
+        mnist_train = self.download_data(self.data_dir)
+
+        self.mnist_train, self.mnist_val = random_split(
+            mnist_train, [55000, 5000])
+
+    def train_dataloader(self):
+        return DataLoader(self.mnist_train, batch_size=int(self.batch_size))
+
+    def val_dataloader(self):
+        return DataLoader(self.mnist_val, batch_size=int(self.batch_size))
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
+        return optimizer
+
+
+def train_mnist(config):
+    model = LightningMNISTClassifier(config)
+    trainer = pl.Trainer(max_epochs=10, show_progress_bar=False)
+
+    trainer.fit(model)
+# __lightning_end__
+
+
+# __tune_callback_begin__
+class TuneReportCallback(Callback):
+    def on_validation_end(self, trainer, pl_module):
+        tune.report(
+            loss=trainer.callback_metrics["avg_val_loss"],
+            mean_accuracy=trainer.callback_metrics["avg_val_accuracy"])
+# __tune_callback_end__
+
+
+# __tune_train_begin__
+def train_mnist_tune(config):
+    model = LightningMNISTClassifier(config, config["data_dir"])
+    trainer = pl.Trainer(
+        max_epochs=10,
+        progress_bar_refresh_rate=0,
+        callbacks=[TuneReportCallback()])
+
+    trainer.fit(model)
+# __tune_train_end__
+
+
+# __tune_checkpoint_callback_begin__
+class CheckpointCallback(Callback):
+    def on_validation_end(self, trainer, pl_module):
+        path = tune.make_checkpoint_dir(trainer.global_step)
+        trainer.save_checkpoint(os.path.join(path, "checkpoint"))
+        tune.save_checkpoint(path)
+# __tune_checkpoint_callback_end__
+
+
+# __tune_train_checkpoint_begin__
+def train_mnist_tune_checkpoint(config, checkpoint=None):
+    trainer = pl.Trainer(
+        max_epochs=10,
+        progress_bar_refresh_rate=0,
+        callbacks=[CheckpointCallback(),
+                   TuneReportCallback()])
+    if checkpoint:
+        # Currently, this leads to errors:
+        # model = LightningMNISTClassifier.load_from_checkpoint(
+        #     os.path.join(checkpoint, "checkpoint"))
+        # Workaround:
+        ckpt = pl_load(
+            os.path.join(checkpoint, "checkpoint"),
+            map_location=lambda storage, loc: storage)
+        model = LightningMNISTClassifier._load_model_state(ckpt, config=config)
+        trainer.current_epoch = ckpt["epoch"]
```

Comment:
this is brittle right? should we raise an issue on their side to address this?

Reply:
Yes, I'll do that tomorrow, didn't get around to it last time